[CL2-6838] i2: [6870] Configure sign-up button (custom link)

alexander-cit opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

alexander-cit committed:
[CL2-6870] Configure sign-up button (custom link)

brentguf commented on front/app/modules/commercial/customizable_homepage_banner/admin/CallToActionButtonSettings.tsx:
```diff
@@ -14,39 +14,62 @@ import {
+  options: (CTASignedOutButton | CTASignedInButton)[];
   currentValue: string;
   signInState: 'not_signed_in' | 'signed_in';
+  labelMessage: MessageDescriptor;
```
It's not needed to pass this as a prop. You can do messages.x directly based on the signInState :-)

alexander-cit replied:
But it will require using this pattern. And for me, it looks a bit ugly

```js
        {
          signed_in: messages.signed_in,
          not_signed_in: messages.not_signed_in,
        }[signInState]
```

brentguf replied:
Haha, how does this beauty look ugly? I think it's a super elegant way of using ifs. And it's used a lot in the front-end, so you better start liking it. 😁 

Anyway, since there are only two possible states (signed in and signed out), you could use a ternary operator as well: `signInState === 'signed_in' ? messages.signed_in : messages.signed_out`

alexander-cit replied:
> so you better start liking it

:smile: 

Ok, will change it.

brentguf commented on front/app/modules/commercial/customizable_homepage_banner/admin/CallToActionButtonSettings.tsx:
```diff
@@ -14,39 +14,62 @@ import {
 
 interface SigninStateProps {
-  options: string[];
+  options: (CTASignedOutButton | CTASignedInButton)[];
```
You can make [conditional props](https://www.benmvp.com/blog/conditional-react-props-typescript/). So if signInState is a certain value, you want CTASignedOutButton[], otherwise you want CTASignedInButton[]

alexander-cit replied:
Great hint! Thank you!

brentguf commented on front/app/modules/commercial/customizable_homepage_banner/admin/CallToActionButtonSettings.tsx:
```diff
@@ -14,39 +14,62 @@ import {
+
+const CTASettingsSignedOut = ({ currentValue }) => (
+  <CallToActionSettingsSigninState
+    options={CALL_TO_ACTION_NOT_SIGNED_IN_OPTIONS}
```
I would also change this one to SIGNED_OUT then?

alexander-cit replied:
Sure, will change it in all the places.

brentguf commented on front/app/modules/commercial/customizable_homepage_banner/admin/CallToActionButtonSettings.tsx:
```diff
@@ -14,39 +14,62 @@ import {
+  labelMessage: MessageDescriptor;
 }
 
 const CallToActionSettingsSigninState = ({
```
I would rename this component. It's hard to tell from the component name what it does. And it should also be extracted in a separate file ideally.

alexander-cit replied:
> And it should also be extracted in a separate file ideally

Sure, will split all files by components a bit later when it's clear how the state can be managed.

> I would rename this component.

I agree. I don't like the word `state` because it's confusingly similar to react state. What do you think about `CTASettingsBySigninStatus`?

brentguf replied:
Or you put the SectionField and Label stuff in the signed in/signed out components (sometimes a little repetition is ok, especially because it's not going to be more than signed in and out. If it is, we can still optimize later if it's needed).

Then the reused part is just the radio buttons and you can call it just that SettingRadios or SettingRadioButtons. Another benefit: we have two fewer props here

alexander-cit replied:
For me, both variants (my initial one and the one you propose) are both fine with their own advantages and disadvantages. And maybe this discussion and this refactoring is a bit (just a bit :smile: ) waste of time, but sure, I will change it :slightly_smiling_face:

brentguf replied:
Haha, didn't need to refactor, but it helped with choosing a name 😄

brentguf commented on front/app/modules/commercial/customizable_homepage_banner/admin/CallToActionButtonSettings.tsx:
```diff
@@ -78,15 +101,11 @@ const CallToActionButtonSettings = ({
-      <CallToActionSettingsSigninState
-        options={CALL_TO_ACTION_NOT_SIGNED_IN_OPTIONS}
+      <CTASettingsSignedOut
         currentValue={call_to_action_not_signed_in_selected_option}
```
You still need to work on this but: JS doesn't use snake_case for variable names. camelCase is used normally.

alexander-cit replied:
But this value of `call_to_action_not_signed_in_selected_option` comes from the server as field of `app_configuration`.

And here it's convenient to use the same name for variable because I can use object desctructuring

```js
  const {
    call_to_action_not_signed_in_selected_option,
    call_to_action_signed_in_selected_option,
  } = customizable_homepage_banner;
 ```
 Otherwise, I would need to do
 ```js
   const callToActionNotSignedInSelectedOption = customizable_homepage_banner.call_to_action_not_signed_in_selected_option;
   const callToActionSignedInSelectedOption = customizable_homepage_banner.call_to_action_signed_in_selected_option;
```
Do you think it looks better?

brentguf replied:
I think you probably won't need to because you'll need to pass a variable, not the setting value. But to answer: the second option is more common, or you could rename after destructuring:

```
  const {
    call_to_action_not_signed_in_selected_option: callToActionNotSignedInSelectedOption ,
    call_to_action_signed_in_selected_option: callToActionSignedInSelectedOption ,
  } = customizable_homepage_banner;
```

It's also going to help if you use `cta` as you proposed 😁

alexander-cit replied:
Not that DRY yet, but yes, looks better :smile:  Will fix it.

brentguf reviewed (commented):
Nothing blocking. :-)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

luucvanderzee committed:
_

luucvanderzee committed:
Merge branch 'CL2-6719-add-and-edit-pages' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6719-add-and-edit-pages

Commit:
Translations updated by CI (extract-intl)

brentguf committed:
Smooth scaling of preview image back-office

brentguf committed:
Move layout heights into module

brentguf committed:
Merge branch 'CL2-6874/homepage-banner-layouts' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6874/homepage-banner-layouts

brentguf committed:
Add feature flag to layout setting

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Danish)

brentguf committed:
Don't use state insertion mechanism for layout heights

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (German)

brentguf committed:
Define layout heights in HeaderImageDropzone (for now)

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6874/homepage-banner-layouts

alexander-cit committed:
[NO TICKET] fix: skip SSRF protection downloading text images on localhost

Commit:
Translations updated by CI (extract-intl)

brentguf committed:
Changelog

brentguf committed:
Merge branch 'CL2-6874/homepage-banner-layouts' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6874/homepage-banner-layouts

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
Update source file en.json

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
Update source file en.json

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Spanish)

luucvanderzee committed:
_

alexander-cit committed:
[NO TICKET] Remove incorrect README

alexander-cit committed:
Merge pull request #1234 from CitizenLabDotCo/alexR-remove-incorrect-readme

[NO TICKET] Remove incorrect README

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.yml (French)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Dutch)

luucvanderzee committed:
Navbar name for page now only editable if page is in navbar

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.yml (French)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.yml (French)

luucvanderzee committed:
Fix: unnecessary requests to insights view input count endpoint

luucvanderzee committed:
Fix: useInsightsInputsCount should update if viewId changes but not queryParameters

luucvanderzee committed:
PageEditor: don't hide title field for about page

jinjagit committed:
Add test - failing as expected

Currently failing as we are not yet returninga relative path to
the respective Idea, but rather an absolute URL = the bug.

luucvanderzee committed:
Merge pull request #1236 from CitizenLabDotCo/fix-useInsightsInputsCount-request-issue

Fix: unnecessary requests to insights view input count endpoint

alexander-cit committed:
Merge branch 'CL2-6874/homepage-banner-layouts' into l10n_CL2-6874/homepage-banner-layouts

alexander-cit committed:
Merge pull request #1229 from CitizenLabDotCo/l10n_CL2-6874/homepage-banner-layouts

New Crowdin updates

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.yml (Dutch)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Dutch)

luucvanderzee committed:
Truncating too long navbar titles

luucvanderzee committed:
Make page body input fields same width as page

jinjagit committed:
Rough fix - BE now passes flaggable_path, not _url, to FE

Test commented out to see what else fails on CI.
Probably now includes some redundant code, to be removed.

jinjagit committed:
Set test to red

Weirdly didn't appear to fail CI when first pushed, so let's
check it does now.

luucvanderzee committed:
Remove secret pages-page

luucvanderzee committed:
Merge branch 'master' into CL2-6719-add-and-edit-pages

luucvanderzee committed:
Update CHANGELOG

luucvanderzee committed:
Merge branch 'CL2-6719-add-and-edit-pages' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6719-add-and-edit-pages

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
Fix failing tests

luucvanderzee committed:
Merge branch 'CL2-6719-add-and-edit-pages' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6719-add-and-edit-pages

luucvanderzee committed:
_

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.yml (Dutch)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Greenlandic)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

jinjagit committed:
Remove redundant code + set test to green

Removed flaggable_url from InappropriateContentFlaggedSerializer.
Removed flaggable_url type definition from notifications.d.ts.
Used flaggable_path in test = green.

kogre committed:
Update source file en.json

jinjagit committed:
Update changelog

jinjagit committed:
Fix unrelated typo that affected emails re. flagged content

The mis-spelling of flaggable_author_name probably meant that
either the value was never displayed on emails regarding flagged
content, or possibly that the email would not be created (failed).

alexander-cit committed:
[CL2-6874] Move custom_style module to core

https://citizenlabco.slack.com/archives/C02K3J4L27N/p1640099906008700

jinjagit committed:
Obey rubocop

luucvanderzee committed:
Fix: navbar title of pages now persisted if removed and added to navbar

jinjagit committed:
Use string, not variable, in test assertion

jinjagit committed:
Return path with leading forward slash from def model_to_path

This removes need to add the forward slash in the FE, when making
use of the return value(s) of this method.

jinjagit committed:
Merge pull request #1237 from CitizenLabDotCo/CL2-6900-fix-url-generation-from-inappropriate-content-notif

CL2-6900 Fix URL generation from inappropriate content notification

luucvanderzee committed:
Page forms now have better error message if slug is already taken

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Dutch)

luucvanderzee committed:
New page form now also includes slug field

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
Slug field now doesn't give error if slug of edited page is unchanged

luucvanderzee committed:
usePageSlugs: fix returning unsubscribe function in wrong block scope

jinjagit committed:
Update changelog

luucvanderzee committed:
Update CHANGELOG

luucvanderzee committed:
Improve slugs of adding/editing pages/navbar items

luucvanderzee committed:
Revert temporary change to text_image_uploader.rb

luucvanderzee committed:
FileUploadField: replace uuid by something simpler

luucvanderzee committed:
Forms that do multiple requests: use Promise.all

luucvanderzee committed:
Fix refetching files after update page

luucvanderzee committed:
Merge pull request #1227 from CitizenLabDotCo/CL2-6719-add-and-edit-pages

CL2-6719 Add and edit pages

luucvanderzee committed:
Merge pull request #1226 from CitizenLabDotCo/crowdin_master

New Crowdin updates

luucvanderzee committed:
Update CHANGELOG

luucvanderzee committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Maori)

luucvanderzee committed:
Merge branch 'master' into l10n_CL2-6719-add-and-edit-pages

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

luucvanderzee committed:
Merge pull request #1228 from CitizenLabDotCo/l10n_CL2-6719-add-and-edit-pages

New Crowdin updates

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

luucvanderzee committed:
Merge branch 'master' into crowdin_master

luucvanderzee committed:
Merge pull request #1240 from CitizenLabDotCo/crowdin_master

New Crowdin updates

alexander-cit committed:
Merge branch 'master' into CL2-6874/homepage-banner-layouts

Commit:
Translations updated by CI (extract-intl)

alexander-cit committed:
[NO TICKET] Resolve translations conflict

luucvanderzee committed:
Merge pull request #1241 from CitizenLabDotCo/alexR-resolve-conflict

[NO TICKET] Resolve translations conflict

alexander-cit committed:
Merge branch 'master' into CL2-6874/homepage-banner-layouts

Commit:
Translations updated by CI (extract-intl)

alexander-cit committed:
Fix ruby code style

alexander-cit committed:
[CL2-6874] Update changelog

alexander-cit committed:
Merge pull request #1077 from CitizenLabDotCo/CL2-6874/homepage-banner-layouts

Cl2 6874/homepage banner layouts

luucvanderzee committed:
Upgrade component library and link to new package

luucvanderzee committed:
Add required ids to Radio components

luucvanderzee committed:
Merge branch 'master' into upgrade-component-library

luucvanderzee committed:
Merge pull request #1244 from CitizenLabDotCo/upgrade-component-library

[NO TICKET] Upgrade component library

luucvanderzee committed:
Upgrade jest and related dependencies

luucvanderzee committed:
Merge pull request #1245 from CitizenLabDotCo/upgrade-jest-and-related-packages

Upgrade jest and related dependencies

alexander-cit committed:
tmp

alexander-cit committed:
tmp2

alexander-cit committed:
tmp2

alexander-cit commented:
Github went crazy!
